src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -35,6 +37,12 @@ def name(self):
 
     def find_matches(self):
         # type: () -> Sequence[Candidate]
+        if self.name in self._factory._constraints:
+            raise InstallationError(
+                "Could not satisfy constraints for '{}': "
+                "installation from path or url cannot be "
+                "constrained to a version".format(self.name)
+            )
```

Comment:
Is this to match the legacy resolver’s behaviour? Because to me it makes sense to allow this (if the version satisfies the constraints, of course).

Reply:
Yes, it's to match the legacy resolver.

I don't mind changing this, but I'd rather not do it in either of these two PRs. As a behaviour change it deserves its own PR (and probably even a news item!)

Reply:
Makes sense to me, this would be something to tell the world about 📰